### PR TITLE
Identity V3: Change user password

### PR DESCRIPTION
**Pull request by zhangzujian**
For #862

API reference:
[Identity API v3 - Change password for user](https://developer.openstack.org/api-ref/identity/v3/index.html#change-password-for-user)

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:
https://github.com/openstack/keystone/blob/a6adc72e3e7ff4ff0bff0702e69ebd8f697d6261/keystone/identity/routers.py#L31
https://github.com/openstack/keystone/blob/a6adc72e3e7ff4ff0bff0702e69ebd8f697d6261/keystone/identity/core.py#L1381
https://github.com/openstack/python-keystoneclient/blob/40642d597deecce4dd81428449353bbd0c2ad5be/keystoneclient/v3/users.py#L203

### CODE REVIEW

**Comment by zhangzujian** [comment]
Ready for review.

File: openstack/identity/v3/users/requests.go [hunk 1]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
 	return
 }
 
+// ChpwdOptsBuilder allows extensions to add additional parameters to
+// the ChangePassword request.
+type ChpwdOptsBuilder interface {
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: Let's called this `ChangePasswordOptsBuilder`. It's a little long, but more clear.

File: openstack/identity/v3/users/requests.go [hunk 2]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
...
+	ToUserChpwdMap() (map[string]interface{}, error)
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: `ToUserChangePasswordMap`

File: openstack/identity/v3/users/requests.go [hunk 3]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
...
+}
+
+// ChpwdOpts provides options for changing password for a user.
+type ChpwdOpts struct {
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: `ChangePasswordOpts`

File: openstack/identity/v3/users/requests.go [hunk 4]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
...
+	// OriginalPassword is the original password of the user.
+	OriginalPassword string `json:"original_password,omitempty"`
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: It looks like the Keystone API will try to authenticate with the original password:

https://github.com/openstack/keystone/blob/a6adc72e3e7ff4ff0bff0702e69ebd8f697d6261/keystone/identity/core.py#L1387

So wouldn't that mean `OriginalPassword` is required? If so, this field should be:

```go
OriginalPassword string `json:"original_password" required:"true"`
```

File: openstack/identity/v3/users/requests.go [hunk 5]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
...
+
+	// Password is the new password of the user.
+	Password string `json:"password,omitempty"`
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: Is `Password` another required field? What happens if this isn't specified?

File: openstack/identity/v3/users/requests.go [hunk 6]
```diff
@@ -204,6 +204,45 @@ func Update(client *gophercloud.ServiceClient, userID string, opts UpdateOptsBui
...
+}
+
+// ToUserChpwdMap formats a ChpwdOpts into a ChangePassword request.
+func (opts ChpwdOpts) ToUserChpwdMap() (map[string]interface{}, error) {
```
Comment (jtopjian) [inline openstack/identity/v3/users/requests.go]: `ToUserChangePasswordMap`

File: openstack/identity/v3/users/results.go [hunk 7]
```diff
@@ -98,6 +98,12 @@ type UpdateResult struct {
 	userResult
 }
 
+// ChpwdResult is the response from a ChangePassword operation. Call its
+// ExtractErr method to determine if the request succeeded or failed.
+type ChpwdResult struct {
```
Comment (jtopjian) [inline openstack/identity/v3/users/results.go]: `ChangePasswordResult`

**Comment by zhangzujian** [comment]
@jtopjian 

I tried the keystone API, and it's OK that values of _password_ and _original_password_ are empty strings. So I removed the JSON tag _omitempty_.

How do you think about the change?

**Comment by zhangzujian** [comment]
Ready for review.

**Comment by zhangzujian** [comment]
check

**Comment by jtopjian** [comment]
Looks good to me. Thank you 😄

**Pull request merged**